src/main/java/com/google/devtools/build/lib/metrics/MetricsCollector.java

```diff
@@ -322,17 +353,63 @@ private TimingMetrics finishTimingMetrics() {
     return timingMetrics.build();
   }
 
+  /**
+   * Millisecond based duration adder which maintain initially zero long sum of milliseconds.
+   * May be accessed from multiple threads.
+   * As it maintains duration as long milliseconds nanoseconds will be ignored.
+   *
+   * The value can be marked invalid with {@code invalidate()}.
+   * For example if the system receives some actions with unknown duration the value is getting unknown (invalid).
+   */
+  static class DurationAdder {
+
+    final LongAdder duration;
+    final AtomicBoolean valid;
+
+    public DurationAdder() {
+      valid = new AtomicBoolean(true);
+      duration = new LongAdder();
+    }
+
+    public void add(Duration duration) {
+      if (valid.get())
+        this.duration.add(duration.toMillis());
```

Comment:
@larsrc-google What do You think about the precision loss here? Wouldn't it be better to switch to microseconds or even nanoseconds here? Initially I was concerned that it might be overflown but maxlong is over 100 days in nanoseconds which should be enough for builds. On the other hand I can imagine some massively parallelized software testing farm

Reply:
Microseconds might be an idea, I know of a few cases where time gets down in the low millis. Nanoseconds would be useless for anything I can think of. Just the fact that we're switching to a different process, let alone creating one, will take microseconds.

Reply:
My bad, I was wrong. It is almost 300 years actually.